### ISSUE TITLE
Template Specs ignore UI definition?

### ISSUE THREAD

**Issue by JustinGrote** [Azure/template-specs#58]
@alex-frankel not sure if I'm missing something here:

Published a template spec using set-aztemplatespec and the uidefinitionfile parameter. Debug shows it loaded fine:

```
Body:
{
  "location": "westus2",
  "tags": {},
  "properties": {
    "uiFormDefinition": {
      "$schema": "https://schema.management.azure.com/schemas/0.1.2-preview/CreateUIDefinition.MultiVm.json#",
      "handler": "Microsoft.Azure.CreateUIDef",
      "version": "0.1.2-preview",
      "parameters": {
        "resourceTypes": [
```
etc...

However when I go to deploy the template spec, it completely ignores it, it seems.
![image](https://user-images.githubusercontent.com/15258962/125835275-dcf955e1-69c0-4eeb-90a8-d0a8a4953869.png)

Am I missing something? This is supported right?

**Comment by JustinGrote** [comment]
I should also note that the UI definition is still present too in Get-TemplateSpec
![image](https://user-images.githubusercontent.com/15258962/125835592-8d4a4268-beaa-4d2c-be34-e2484c73ddee.png)

**Comment by JustinGrote** [comment]
I tried it with the traefik azure quickstart UI definition which I know works and still no joy, the "deploy to azure" button works fine tho.
![image](https://user-images.githubusercontent.com/15258962/125855492-fdb5d6fe-cc29-4551-8d6c-19d9a2b8e37c.png)

**Comment by alex-frankel** [comment]
This was answered on twitter, but want to close the loop here. We built support for this into the templateSpec API, but have not finished the portal support to read this UiDefinition if it is provided. Our current ETA is that this will be completed in about a month. cc @apclouds

**Comment by peter-bertok** [comment]
> 
> 
> This was answered on twitter, but want to close the loop here. We built support for this into the templateSpec API, but have not finished the portal support to read this UiDefinition if it is provided. Our current ETA is that this will be completed in about a month. cc @apclouds

I'm eagerly awaiting the release of this. It's the #1 most requested feature among our customers. Most administrators prefer to have a nice GUI they can click through, and as a consultant I haven't been able to deliver anything for them because this was missing.

**Comment by Bad-Rog** [comment]
I'd like to second peter-bertoks comment...
This will be a great feature, please update us here, once it is implemented.

**Comment by peter-bertok** [comment]
> Our current ETA is that this will be completed in about a month

Is there any update on the progress of this feature?

**Comment by Bad-Rog** [comment]
> This was answered on twitter, but want to close the loop here. We built support for this into the templateSpec API, but have not finished the portal support to read this UiDefinition if it is provided. Our current ETA is that this will be completed in about a month. cc @apclouds

Guys, can we get an updated on this Issue?

**Comment by alex-frankel** [comment]
Just received an ETA that portal support for custom UiDefinition will be available at the end of September.

**Comment by JustinGrote** [comment]
@alex-frankel later than hoped but thanks for the update all the same!

**Comment by peter-bertok** [comment]
> portal support for custom UiDefinition will be available at the end of September

@alex-frankel will it be available in the preview portal before then for testing?

**Comment by alex-frankel** [comment]
More than likely, but I don't know for sure at this point. As soon as it's publicly available for use, we will let this thread know.

**Comment by peter-bertok** [comment]
It appears to be enabled in the Preview portal now, but I get a generic error:

![image](https://user-images.githubusercontent.com/14988123/133548681-ee2f5f16-82f4-46be-bf3e-03b20cbea905.png)

Is this because the feature is not ready, or because there's a new schema for the UI definitions used by Template Specs?

**Comment by alex-frankel** [comment]
Yes, the schema is being tweaked slightly. The docs are not yet available, but if you are very eager you can take a look at the PR for what the schema will be:
https://github.com/Azure/azure-resource-manager-schemas/pull/1906

**Comment by Bad-Rog** [comment]
The feature seems to be implemented into Production Portal now,  however i still get the above mentioned error...
By when will there be a documentation to know what to amend?

thanks a lot

**Comment by alex-frankel** [comment]
The goal is to have docs available by the end of the month

**Comment by peter-bertok** [comment]
I managed to get it working:

![image](https://user-images.githubusercontent.com/14988123/134264170-00ce1e58-ac75-4b46-a23e-4d0088542b2e.png)

One thing I noticed is that it still shows the "Azure Marketplace Terms" acceptance form, which shouldn't be there for Template Specs.

Some other features I'd love to see are:

- Previews of what resources will be deployed, vaguely like "-WhatIf", but ideally with a GUI look and feel. E.g.: the correct icons, etc...
- The ability to customise the icon of the template spec. In this example I would like it to have the Azure DNS Zone icon instead of the generic template icon.

**Comment by alex-frankel** [comment]
>One thing I noticed is that it still shows the "Azure Marketplace Terms" acceptance form, which shouldn't be there for Template Specs.

My understanding is this is shown for all custom template deployments. It should also be there if you click a "Deploy to Azure" button in GitHub for example.

Definitely would be great to add What-if output (nicely visualized) to this experience.

Custom icons for TemplateSpecs is also an interesting idea. Do you mind making a separate issue to keep track of that?

**Comment by alex-frankel** [comment]
Here is the doc:
https://docs.microsoft.com/azure/azure-resource-manager/templates/template-specs-create-portal-forms

Going to close this for now. Let us know if you run into any issues.

**Comment by RiverHeart** [comment]
Hey guys, sorry for the necro but can someone comment on whether the ARM syntax that populates in the form actually resolves correctly when applied? So, if we have a resource group in "Central US" the following will resolve properly.

`Location: [resourceGroup().location]`

**Comment by alex-frankel** [comment]
If you have a question specific to UI Form Definition, I would open an issue here:
https://github.com/Azure/portaldocs
